### PR TITLE
Report git tag as fw version

### PR DESCRIPTION
**Pull request by dl1com**
Implements changes to report the firmware version in cnfInfo according to the current git tag.

- The git tag has to be in the format x.y.z, where x,y,z are 1-3 digits. (as defined by semver.org).
- In case the git tag does not match this format, 0.0.0 is used.
- In case there have been changes since the tag was applied (indicated by an appendix in the version string by `git describe --tags`, for example "1.0.0-4-gc5abe8c", the "dirty" flag is set to indicate that the software does as been altered.

Still to do (any help appreciated!):

- [x] Fix firmware unit tests (dynamically generated version.h is missing in case of running unit tests standalone)

- [x] Update the APIv6 spec in ayab-manual. *cnfInfo* consists now of 6 bytes. PR: https://github.com/AllYarnsAreBeautiful/ayab-manual/pull/27

```
payload[0] = cnfInfo_msgid;
payload[1] = API_VERSION;
payload[2] = FW_VERSION_MAJ;
payload[3] = FW_VERSION_MIN;
payload[4] = FW_VERSION_PATCH;
payload[5] = FW_VERSION_DIRTY;
```

- [x] Update the relevant code in ayab-desktop to parse the FW_VERSION_PATCH and FW_VERSION_DIRTY flags when receiving cnfInfo, PR: https://github.com/AllYarnsAreBeautiful/ayab-desktop/pull/550
- [x] Test whether the preBuild.py runs properly on Linux, Windows and Mac

### CODE REVIEW

**Comment by dl1com** [comment]
This PR and the related ones are feature complete.
Before merging, I'd like to ask somebody using PlatformIO on Windows and MacOS to test if the build runs successfully on their system, as preBuild.py is now called.
/cc @VIPQualityPost @t0mpr1c3

**Comment by VIPQualityPost** [comment]
Just build, then try to start AYAB and check the logs to see if everything went smoothly? Or is there more to check?

**Comment by t0mpr1c3** [comment]
I can test Windows.

**Comment by VIPQualityPost** [comment]
Builds fine, but the version does not seem to get picked up correctly (mac):
```
23-07-07 10:33:03 root     DEBUG    Entered state CONFIGURING
23-07-07 10:33:07 root     DEBUG    Entered state CHECKING
23-07-07 10:33:07 Engine   DEBUG    {'portname': '/dev/cu.usbmodem1401', 'machine': 'KH900_KH930_KH940_KH965', 'mode': 'SINGLEBED', 'num_colors': 2, 'start_row': 0, 'inf_repeat': False, 'start_needle': 84, 'stop_needle': 116, 'alignment': 'CENTER', 'auto_mirror': False, 'continuous_reporting': False}
23-07-07 10:33:07 root     DEBUG    Entered state KNITTING
23-07-07 10:33:07 Control  DEBUG    State CONNECT
23-07-07 10:33:07 Control  DEBUG    Port name: /dev/cu.usbmodem1401
23-07-07 10:33:07 Control  DEBUG    State VERSION_CHECK
23-07-07 10:33:08 Control  INFO     API v6, FW v0.0
23-07-07 10:33:08 Control  DEBUG    State INIT
23-07-07 10:33:08 root     INFO     Notification: Initializing firmware
23-07-07 10:33:08 Control  DEBUG    State REQUEST_START
23-07-07 10:33:08 root     INFO     Notification: Please start machine. (Set the carriage to mode KC-I or KC-II and move the carriage over the left turn mark).
```

**Comment by dl1com** [comment]
Following the discussion here, I changed the 'dirty' flag into a suffix string of max. 16 characters length.
Please retest.
For testing, you have to locally create a tag in a valid semver format. Expected tags i.e. `0.96.0`, `0.96.0-rc1`, `0.96.0-23-g9a97c14`

Depending PRs:

- https://github.com/AllYarnsAreBeautiful/ayab-manual/pull/28
- https://github.com/AllYarnsAreBeautiful/ayab-desktop/pull/550

**Comment by VIPQualityPost** [comment]
OK- I think it works? I'm not totally sure if I did it right.
New clone of repo -> `git tag -a 0.96.0-23-g9a97c14 -m "test tag"`
Upload OK. Logs:
```
23-07-14 21:46:30 root     DEBUG    Entered state KNITTING
23-07-14 21:46:30 Control  DEBUG    State CONNECT
23-07-14 21:46:30 Control  DEBUG    Port name: /dev/cu.usbmodem11401
23-07-14 21:46:30 Control  DEBUG    State VERSION_CHECK
23-07-14 21:46:31 Control  INFO     API v6, FW v0.96
23-07-14 21:46:31 Control  DEBUG    State INIT
23-07-14 21:46:31 root     INFO     Notification: Initializing firmware
23-07-14 21:46:31 Control  DEBUG    State REQUEST_START
23-07-14 21:46:31 root     INFO     Notification: Please start machine. (Set the carriage to mode KC-I or KC-II and move the carriage over the left turn mark).
```

**Comment by dl1com** [comment]
> OK- I think it works? I'm not totally sure if I did it right. It didn't carry over the "full" firmware numbering but it got the main. I don't know where the "dirty" information is supposed to be revealed? New clone of repo -> `git tag -a 0.96.0-23-g9a97c14 -m "test tag"` Upload OK. Logs:
> 
> ```
> 23-07-14 21:46:30 root     DEBUG    Entered state KNITTING
> 23-07-14 21:46:30 Control  DEBUG    State CONNECT
> 23-07-14 21:46:30 Control  DEBUG    Port name: /dev/cu.usbmodem11401
> 23-07-14 21:46:30 Control  DEBUG    State VERSION_CHECK
> 23-07-14 21:46:31 Control  INFO     API v6, FW v0.96
> 23-07-14 21:46:31 Control  DEBUG    State INIT
> 23-07-14 21:46:31 root     INFO     Notification: Initializing firmware
> 23-07-14 21:46:31 Control  DEBUG    State REQUEST_START
> 23-07-14 21:46:31 root     INFO     Notification: Please start machine. (Set the carriage to mode KC-I or KC-II and move the carriage over the left turn mark).
> ```

Did you use the matching branch of the desktop software for this test? There have been changes as well to enable the suffix information being printed in the log.
https://github.com/AllYarnsAreBeautiful/ayab-desktop/pull/550

The information should be printed in the 5th line of your output (API v6, FW...)

**Comment by VIPQualityPost** [comment]
I built that pr: `gh pr checkout 550` and then built. When I run it I get more firmware info that I got before but still not the full firmware version:
```
(base) mateijordache@mateibook ayab-firmware % git status
HEAD detached at 0.96.0-23-g9a97c14 
```

As reported in AYAB log on building pr 550:
```
2023-07-15 12:03:31,169 Control  DEBUG    State VERSION_CHECK
2023-07-15 12:03:31,907 Control  INFO     API v6, FW v0.96.0
2023-07-15 12:03:31,908 Control  DEBUG    State INIT
```

**Comment by dl1com** [comment]
> I just tried: `gh pr checkout 550` and then build. When I run it I get more firmware info than I got before but still not the full firmware version, it does not report the tag but does get all of the first segment now:
> 
> ```
> (base) me@mycomputer ayab-firmware % git status
> HEAD detached at 0.96.0-23-g9a97c14 
> ```
> 
> As reported in AYAB log on building pr 550:
> 
> ```
> 2023-07-15 12:03:31,169 Control  DEBUG    State VERSION_CHECK
> 2023-07-15 12:03:31,907 Control  INFO     API v6, FW v0.96.0
> 2023-07-15 12:03:31,908 Control  DEBUG    State INIT
> ```

Ok, I'll investigate.

**Comment by dl1com** [comment]
> 
> Ok, I'll investigate.

Just tested again, for me PR#550 together with this one work fine (tested tags: 0.96.0-rc1, 0.96.0-23-g9a97c14)

**Comment by dl1com** [comment]
@VIPQualityPost  @t0mpr1c3 pls approve this PR if you feel comfortable :)

**Comment by VIPQualityPost** [comment]
Yes, I think I'm doing something wrong. When I try to make new tags, the GIT Version doesn't increment, so I think this is user error on my side, but I can't figure out how to get it to take the tag and put that into the build version, although the 'dirty' mark seems to work fine:
```
Pre build script
0.96.0-23-g9a97c14
GIT Version: 0.96.0 dirty:1
```

Maybe @t0mpr1c3 can confirm it works for Windows and then approve?

**Comment by VIPQualityPost** [comment]
I cleaned my ayab-firmware and now it works! So it was just pebcak..

**Review by VIPQualityPost — APPROVED** [review]

**Pull request merged**
